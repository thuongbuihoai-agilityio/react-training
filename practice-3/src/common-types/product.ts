export interface Product {
  id: string;
  name: string;
  price: number;
  images: string[];
  quantity: number;
  categoryId: string;
  description: string;
  product?: {};
}

export interface ProductItemProps {
  product: Product;
  deleteProduct: (id: string) => void;
}

export interface ProductCardProps {
  product: Product;
}

export interface ProductContext {
  products: Product[];
  dispatch: Function,
  searchValue: string;
  setSearchValue: Function;
}