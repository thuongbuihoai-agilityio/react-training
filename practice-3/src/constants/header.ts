const USERS = {
  username: "Hoai Thuong",
};

export { USERS }