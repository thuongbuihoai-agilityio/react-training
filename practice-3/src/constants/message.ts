const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

export { SUCCESS_MSG };
