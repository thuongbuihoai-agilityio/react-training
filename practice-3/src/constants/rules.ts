const RULES = {
  REQUIRED: "required",
  NUMBER: "number",
};

export { RULES }