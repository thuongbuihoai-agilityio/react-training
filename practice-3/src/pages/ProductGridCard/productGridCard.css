.product {
  position: relative;
  margin: 18px;
  z-index: 0;
}

.product,
.product__image {
  width: 250px;
}

.product__image:hover {
  opacity: 0.5;
}

.product__image:hover+.btn__delete {
  display: block;
}

.product__content {
  text-align: center;
}

.product__title {
  font-size: 16px;
  cursor: pointer;
  text-transform: uppercase;
}

.productViewPage__price {
  color: var(--text-primary);
  font-weight: bold;
  margin: 0;
}