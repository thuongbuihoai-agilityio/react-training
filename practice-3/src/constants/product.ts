import { Product } from "types/product";

const PRODUCT_CONTENT: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    "/src/assets/images/products/cheese-cake.jpg",
    "/src/assets/images/products/cheese-cake.jpg",
    "/src/assets/images/products/cheese-cake.jpg"
  ]
}

export { PRODUCT_CONTENT };
