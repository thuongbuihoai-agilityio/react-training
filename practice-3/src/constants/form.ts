import { RULES } from "./rules"

const FORM_VALUES = {
  categoryId: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  name: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  price: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  quantity: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  description: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  images: {
    value: "",
    rules: [],
    error: "",
  }
}

export default FORM_VALUES
