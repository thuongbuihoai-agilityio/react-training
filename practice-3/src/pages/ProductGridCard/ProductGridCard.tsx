import React, { useCallback, useState } from "react";
import { Link } from "react-router-dom";
import Title from "@/components/common/Title/Title";
import Price from "@/components/Price/Price";
import { ProductItemProps } from "@/types/product";
import ModalDelete from "@/components/Modal/ModalDelete/ModalDelete";
import "./productGridCard.css";

const ProductGridCard: React.FC<ProductItemProps> = ({
  product,
  deleteProduct,
}) => {
  const [openModalDelete, setOpenModalDelete] = useState(false);
  const toggleModalUpdate = useCallback(() => {
    setOpenModalDelete(!openModalDelete);
  }, [openModalDelete]);

  return (
    <>
      <div data-testid="product-grid-card" className="product">
        <img className="product__image" src={product.images[0]} />
        <button
          data-testid="open-modal-delete"
          onClick={toggleModalUpdate}
          className="btn btn__delete"
        >
          <i className="fa fa-trash"></i>
        </button>
        <div className="product__content">
          <Link className="productViewPage__link" to={`/product/${product.id}`}>
            <Title className="productViewPage__title" text={product.name} />
          </Link>
          <Price className="productViewPage__price" value={product.price} />
        </div>
        {openModalDelete && (
          <ModalDelete
            id={product.id}
            deleteProduct={deleteProduct}
            hideModalDelete={toggleModalUpdate}
          />
        )}
      </div>
    </>
  );
};

export default ProductGridCard;
