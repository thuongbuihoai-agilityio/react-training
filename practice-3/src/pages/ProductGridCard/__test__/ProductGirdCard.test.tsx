import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { PRODUCT_MOCKING } from "@/constants/product";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import ProductGridCard from "../ProductGridCard";

describe("ViewProductItem component", () => {
  test("should render product grid card component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render (
      <Router location={history.location} navigator={history}>
        <ProductGridCard product={PRODUCT_MOCKING} deleteProduct={() => {}} />
      </Router>
    );
    expect(getByTestId("product-grid-card")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router location={history.location} navigator={history}>
        <ProductGridCard product={PRODUCT_MOCKING} deleteProduct={() => {}} />
      </Router>);
    expect(asFragment()).toMatchSnapshot();
  });
})