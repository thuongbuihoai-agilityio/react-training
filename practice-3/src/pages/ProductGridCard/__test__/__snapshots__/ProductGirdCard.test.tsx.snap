// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewProductItem component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="product"
    data-testid="product-grid-card"
  >
    <img
      class="product__image"
      src="/src/assets/images/products/cheese-cake.jpg"
    />
    <button
      class="btn btn__btn btn__delete"
    >
      <i
        class="fa fa-trash"
      />
    </button>
    <div
      class="product__content"
    >
      <a
        class="productViewPage__link"
        href="/product/1"
      >
        <h2
          class="productViewPage__title"
          data-testid="title-test"
        >
          Cheese cake
        </h2>
      </a>
      <p
        class="productViewPage__price"
        data-testid="price"
      >
        $20
      </p>
    </div>
  </div>
</DocumentFragment>
`;
